### PR TITLE
Adding o16nc interaction channel

### PR DESCRIPTION
**Pull request by SWakely**
o16nc_p.py and o16nc_n.py have been written to model neutral current interactions on O-16 (proton and neutron emission channels). 
The cross-section as a function of energy has been fitted from data presented in (https://link.aps.org/doi/10.1103/PhysRevC.98.034613) and the threshold energy estimated from these fits.
The energy spectrum of the outgoing particle is assumed to be uniform as no information could be found for the correct shape of the spectrum. The outgoing energy is therefore generated at random from the available energy range.

### CODE REVIEW

File: sntools/interaction_channels/o16nc_n.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+e_thr = 19.5  # approximate energy threshold of neutron emission with no gamma emission (MeV)
+mN = 939.7 # neutron mass (MeV)
+'''e_thr_g = e_thr + 6.18  # energy threshold of neutron emission with gamma emission (MeV)'''
+epsilon = 0.001  # for approximating DiracDelta distribution below
+
+# List of neutrino flavors ("e", "eb", "x", "xb") that interact in this channel.
+possible_flavors = ("e", "eb", "x", "xb")
+
+x_sec_data = [(e_thr, None    , 0.0    , None    , None    , None    , None    , None    ,
```
Comment (JostMigenda) [inline sntools/interaction_channels/o16nc_n.py:35]: It looks like you’re going to a lot of effort here and over the next ~60 lines of code to make this extremely general and capable of supporting all emission channels—but only one of them is actually used in the code, unless I’m missing something?
I think I would prefer to remove the code for those other channels here, to make it a lot shorter and easier to understand.

**Review by JostMigenda — COMMENTED** [review]
Many thanks for your contributions @SWakely! At first glance, this looks really good already. I have a few minor comments to clean the code up a bit; we can go over those at our meeting next week. I will take a closer look at the paper next week and double check the physics, too.

Before I merge this, it also needs unit tests and documentation. Let me know if you are able to work on those or if you’d like me to take over and write them!

**Comment by JostMigenda** [comment]
Hi @SWakely! We talked about the `random.random()` call in `get_eE(eNu, cosT)` at our meeting recently and noticed that the current implementation would be a problem for `bounds_eE` since a different random value would be used for the lower and upper bound. (To fix that, we can call `get_eE` just once and then use that single value to calculate the lower and upper bounds.)
I’ve also said I would check if this could lead to problems elsewhere in the code. In the implementation in `channel.py`, there are a few occasions where `get_eE()` gets called indirectly; here are my thoughts about those:

* [In line 40](https://github.com/SWakely/sntools/blob/618b041413bc50f7f60add60b7be4260b7d591a6/sntools/channel.py#L40) & [line 69](https://github.com/SWakely/sntools/blob/618b041413bc50f7f60add60b7be4260b7d591a6/sntools/channel.py#L69), via `bounds_eE`: This is then used as integration range for `dSigma_dE(eNu, eE)`—but since that doesn’t depend on `eE`, having a random value isn’t a problem here as long as the width of the integration range is two epsilon.
* [In line 71](https://github.com/SWakely/sntools/blob/618b041413bc50f7f60add60b7be4260b7d591a6/sntools/channel.py#L71), via `generate_event()`: This is where the value that ends up in the output file is generated. Here, it’s totally fine to have a random value.

In summary: I think once `bounds_eE` is fixed, everything else should be alright.

File: sntools/interaction_channels/o16nc_n.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+        Inputs:
+            eNu: neutrino energy
+            eE:  energy of outgoing (detected) particle
+        """
+        if eNu < e_thr or abs(self.get_eE(eNu) - eE) > epsilon:
+            # This should never happen, since we set bounds for eE and eNu accordingly above
+            # ... but just in case:
+            return 0
+
+        sigma = fit()(eNu)*10**(-42)  # cross-section at eNu from the interp1d fit
```
Comment (JostMigenda) [inline sntools/interaction_channels/o16nc_n.py:92]:
```suggestion
        sigma = fit(eNu)*10**(-42)  # cross-section at eNu from the interp1d fit 
```
Adjust here if `fit` is turned from a function into a variable above.

File: sntools/interaction_channels/o16nc_n.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+        emission angle of the outgoing (detected) particle.
+
+        Input:
+            eNu:  neutrino energy (MeV)
+            cosT: cosine of the angle between neutrino and outgoing (detected) particle
+        """
+        # Energy dependence is unclear, so we use a constant value for now.
+        if abs(cosT) > 1:
+            return 0
+        return 0.5
```
Comment (JostMigenda) [inline sntools/interaction_channels/o16nc_n.py:109]: This should be `0.5 * sigma` (where sigma is defined as in `dSigma_dE` above), to ensure that integrating `dSigma_dCosT` over `cosT` gives the physically correct value.

(Right now, this shouldn’t affect the output of sntools, since that’s only affected by the cosT-dependence and not the absolute value of this function. But I’d rather be physically accurate right away, so this doesn’t cause issues in the future.)

**Review by JostMigenda — CHANGES_REQUESTED** [review]
I just looked through the implementation in more detailed; here are some changes I would suggest. They’re fairly straightforward and shouldn’t take much time. Hope it’s all clear, but let me know if I’ve misunderstood something!

(Note: All comments apply analogously to `o16nc_p.py`.)

**Comment by JostMigenda** [comment]
Added a brief section in the documentation; the only thing that’s missing before this PR can be merged are unit tests for both channels. I’ll put that on my todo list for next week.

**Review by JostMigenda — APPROVED** [review]

**Comment by JostMigenda** [comment]
Added some basic tests for both subchannels, so this is now ready to merge. Once again, @SWakely, thanks for the contributions and congratulations on your first successful PR!
